File: Python/linkedlist.py
from node import Node
class Linked_List:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0
    def push(self, val):
        new_node = Node(val)
        if self.length == 0:
            self.head = new_node
            self.tail = new_node
            self.length += 1
        else:
            temp = self.tail
            temp.next = new_node
            self.tail = new_node
            self.length += 1
    def pop(self):
        if self.length == 0:
            return -1
        elif self.length == 1:
            val = self.tail.val
            self.head = None
            self.tail = None
            length -= 1
            return val
        else:
            cur = self.head
            while cur.next != self.tail:
                cur = cur.next
            value = cur.next.val
            cur.next = None
            self.tail = cur
            length -= 1
            return value
    def reverse(self):
        before = None
        cur = self.head
        after = None
        while cur:
            after = cur.next
            cur.next = before
            before = cur
            cur = after
        self.tail = self.head
        self.head = before
    def is_empty(self):
        return self.length == 0
    def __len__(self):
        return self.length
    def __iter__(self):
        cur = self.head
        while cur:
            yield cur.val
            cur = cur.next
    def __getitem__(self, index):
        if index < 0 or index >= self.length:
            raise IndexError("invalid index")
        else:
            cur = self.head
            for x in range(index):
                cur = cur.next
            return cur.val
    def __str__(self):
        cur = self.head
        string = ""
        while cur:
            string += str(cur.val) + " -> "
            cur = cur.next
        if string:
            string = string[:len(string) - 3]
        else:
            string = "[]"
        return string

File: C++/LinkedList.cpp
#include <iostream>
#include "LinkedList.h"

LinkedList::LinkedList()
{
    head = nullptr;
    tail = nullptr;
    length = 0;
}

LinkedList::~LinkedList()
{
    node* temp = head;
    while(temp != nullptr) {
        node* tem_next = temp->next;
        delete temp;
        temp = tem_next;
    }
}
int LinkedList::size() {
    return length;
}
void LinkedList::push(int val) {
    node* new_node = new node(val);
    if(length == 0) {
        head = new_node;
        tail = new_node;
    } else {
        tail->next = new_node;
        tail = new_node;
    }
    length++;
}
int LinkedList::pop() {
    if(length == 0) {
        std::cout << "error out of index" << std::endl;
        return -1;
    } else if(length == 1) {
        int val = tail->value;
        tail = nullptr;
        head = nullptr;
        return val;
    } else {
        node* temp = head;
        while(temp->next != tail) {
            temp = temp->next;
        }
        int val = temp->next->value;
        tail = temp;
        temp->next = nullptr;
        delete temp->next;
        return val;
    }
}
bool LinkedList::isEmpty() {
    return length == 0;
}
void LinkedList::show() {
    node* temp = head;
    while(temp != nullptr) {
        std::cout << temp->value << "->";
        temp = temp->next;
    }
    std::cout << "" << std::endl;
}
void LinkedList::reverse() {
    if(length == 0) {
        std::cout << "cant be done" << std::endl;
    } else {
        node* cur = head;
        node* before {nullptr};
        node* after {nullptr};
        while(cur != nullptr) {
            after = cur->next;
            cur->next = before;
            before = cur;
            cur = after;
        }
        head = tail;
 
    }
}